cumulusci/robotframework/BaseLibrary.py

```diff
@@ -0,0 +1,42 @@
+import logging
+from robot.libraries.BuiltIn import BuiltIn, RobotNotRunningError
+from cumulusci.robotframework.utils import RetryingSeleniumLibraryMixin
+
+
+class BaseLibrary(RetryingSeleniumLibraryMixin):
+    ROBOT_LIBRARY_SCOPE = "GLOBAL"
+
+    def __init__(self, debug=False):
+        self.debug = debug
+        self.retry_selenium = True
+        # Turn off info logging of all http requests
+        logging.getLogger("requests.packages.urllib3.connectionpool").setLevel(
+            logging.WARN
+        )
+
+    @property
+    def salesforce_api_version(self):
+        try:
+            client = self.cumulusci.tooling
+            response = client._call_salesforce(
+                "GET", "https://{}/services/data".format(client.sf_instance)
+            )
+            latest_api_version = float(response.json()[-1]["version"])
+            return latest_api_version
+
+        except RobotNotRunningError:
+            # not sure if this should return None, a reasonable default value,
+            # or raise an exception. :-\
+            return None
+
+    @property
+    def builtin(self):
+        return BuiltIn()
+
+    @property
+    def cumulusci(self):
+        return self.builtin.get_library_instance("cumulusci.robotframework.CumulusCI")
+
+    @property
+    def salesforce(self):
+        return self.builtin.get_library_instance("cumulusci.robotframework.Salesforce")
```

Comment:
Hmm. The Salesforce library extends BaseLibrary, which means it can get itself via the `salesforce` attribute. I guess that's mostly harmless but it feels a little weird.

Are our product libraries meant to extend BaseLibrary and use the `salesforce` attribute to call salesforce keywords, or are they meant to extend the Salesforce library and inherit its keywords to use directly?

Reply:
yes, it is weird that the salesforce library has a property referring to itself. I'm still pondering the wisdom of that. On the one hand, Salesforce is special in that it has core keywords, but on the other hand, we should drink our own champaign.

As to the other question, at this point they are meant to extend BaseLibrary and use the `salesforce` attribute. I'm open to the idea of inheriting from Salesforce, but I don't think that buys us much at the moment. I'm open to considering it if there's a compelling reason, but it would break a bunch of keywords.

Reply:
Okay. I guess we could override the salesforce attribute on the salesforce library to make it None or something like that.